Neos.Neos/Classes/Controller/Service/AssetProxiesController.php

```diff
@@ -0,0 +1,167 @@
+ * source code.
+ */
+
+use Neos\Flow\Annotations as Flow;
+use Neos\Flow\Mvc\Controller\ActionController;
+use Neos\Flow\Mvc\Exception\StopActionException;
+use Neos\Flow\Mvc\Exception\UnsupportedRequestTypeException;
+use Neos\Flow\Mvc\View\ViewInterface;
+use Neos\FluidAdaptor\View\TemplateView;
+use Neos\Media\Domain\Repository\AssetRepository;
+use Neos\Media\Domain\Service\AssetSourceService;
+use Neos\Media\Domain\Repository\TagRepository;
+use Neos\Media\Exception\AssetSourceServiceException;
+use Neos\Neos\Controller\BackendUserTranslationTrait;
+use Neos\Neos\View\Service\AssetJsonView;
+
+/**
+ * Rudimentary REST service for assets
+ *
+ * @Flow\Scope("singleton")
+ */
+class AssetProxiesController extends ActionController
+{
+    use BackendUserTranslationTrait;
+
+    /**
+     * @Flow\Inject
+     * @var AssetRepository
+     */
+    protected $assetRepository;
+
+    /**
+     * @Flow\Inject
+     * @var AssetSourceService
+     */
+    protected $assetSourceService;
+
+    /**
+     * @Flow\Inject
+     * @var TagRepository
+     */
+    protected $tagRepository;
+
+    /**
+     * @Flow\InjectConfiguration(package="Neos.Media", path="asyncThumbnails")
+     * @var boolean
+     */
+    protected $asyncThumbnails;
+
+    /**
+     * @var array
+     */
+    protected $viewFormatToObjectNameMap = array(
+        'html' => TemplateView::class,
+        'json' => AssetJsonView::class
+    );
+
+    /**
+     * A list of IANA media types which are supported by this controller
+     *
+     * @var array
+     * @see http://www.iana.org/assignments/media-types/index.html
+     */
+    protected $supportedMediaTypes = array(
+        'text/html',
+        'application/json'
+    );
+
+    /**
+     * @param ViewInterface $view
+     * @return void
+     */
+    public function initializeView(ViewInterface $view)
+    {
+        $view->assign('asyncThumbnails', $this->asyncThumbnails);
+    }
+
+    /**
+     * Shows a list of assets
+     *
+     * @param string $searchTerm An optional search term used for filtering the list of assets
+     * @param string $assetSourceIdentifier If specified, results are only from the given asset source
+     * @param int $limitPerAssetSource The maximum number of results shown per asset source
+     * @return void
+     */
+    public function indexAction(string $searchTerm = '', string $assetSourceIdentifier = '', int $limitPerAssetSource = 10): void
+    {
+        $assetSources = $this->assetSourceService->getAssetsSources();
+        $assetProxies = [];
+        foreach ($assetSources as $assetSource) {
+            if ($assetSourceIdentifier !== '' && $assetSource->getIdentifier() !== $assetSourceIdentifier) {
+                continue;
+            }
+            try {
+                $assetProxyRepository = $assetSource->getAssetProxyRepository();
+                $assetProxyQueryResult = $assetProxyRepository->findBySearchTerm($searchTerm);
```

Comment:
I guess this one needs a limit, too!

Reply:
Well, that's implemented in line 108.

Reply:
What I meant is: We need to limit that on the asset source's side. Otherwise an asset source could download all images from Wikipedia only to be limited afterwards :)

Reply:
IMO, that is the responsibility of the asset source. The interface says that it should return a query result with Iterator support. What I do in my asset source is, that I fetch x asset proxies at a time and if you iterate further, fetch the next ones. Not all asset sources can support something like limit or offset, because they simply have a different way of paginating results.